### PR TITLE
Adding config to only allow manual trigger if previous stage was success

### PR DESCRIPTION
**Pull request by kritika-singh3**
Issue: #6294 

 - [x]  Add config migration to `126`
 - [ ] Bump up Pipeline Config API to v9 to support `allowOnlyOnSuccess`

### CODE REVIEW

**Comment by arvindsv** [comment]
Should we also disable the ">>>" button (manual trigger button) in stage history and stage details?

**Comment by kritika-singh3** [comment]
> Should we also disable the ">>>" button (manual trigger button) in stage history and stage details?

I was of the opinion of keeping it active and adding a server health message. 
If we disable it, we would need to inform the user of the reason behind it. WDYT?

**Comment by arvindsv** [comment]
Since it's an action, the corresponding error should ideally show somewhere near the action, as a result. Right? It shouldn't feel like the action completed. It's easy to miss an addition to an existing set of errors. If it was "3 errors" and it becomes "4 errors", you'd never notice.

Typically, I think of using the server health messages to provide information about things happening in the background.

Also: When a person doesn't have permissions to trigger the next stage, the manual trigger gate shows a message on mouse-over. So, it's not out of the ordinary. If we're allowing them to click it, then an error should show on the page itself, once the error happens.

File: server/webapp/WEB-INF/rails/app/views/admin/stages/new.html.erb [hunk 1]
```diff
@@ -1,65 +1,89 @@
...
-                            <label for="auto">On Success</label>
-                        </div>
+      <div class="fieldset">
+        <div class="form_item required">
+          <%= label_tag stage_name_id = random_dom_id("stage_name_"), required_label_text('Stage Name') %>
+          <%= f.text_field com.thoughtworks.go.config.StageConfig::NAME, :class => "form_input", :id => stage_name_id -%>
+          <%= error_message_on(@stage, "name", :css_class => "form_error") %>
+        </div>
+        <div class="form_item checkbox_row">
+          <label>Trigger Type in new:</label>
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply server/webapp/WEB-INF/rails/app/views/admin/stages/new.html.erb]: No. Reverted it.

File: server/src/main/java/com/thoughtworks/go/server/service/PipelineHistoryService.java [hunk 2]
```diff
@@ -199,11 +199,13 @@ private void appendFollowingStagesFromConfig(String pipelineName, StageInstanceM
...
     private void populateCanRunStatus(Username username, PipelineInstanceModel pipelineInstanceModel) {
         for (StageInstanceModel stageHistoryItem : pipelineInstanceModel.getStageHistory()) {
+            ServerHealthStateOperationResult result = new ServerHealthStateOperationResult();
             boolean canRun = scheduleService.canRun(
                     pipelineInstanceModel.getPipelineIdentifier(), stageHistoryItem.getName(),
                     CaseInsensitiveString.str(username.getUsername()), pipelineInstanceModel.hasPreviousStageBeenScheduled(
-                            stageHistoryItem.getName()));
+                            stageHistoryItem.getName()), result);
             stageHistoryItem.setCanRun(canRun);
+            stageHistoryItem.setErrorMessage(canRun || result.getServerHealthState() == null ? "" : result.getServerHealthState().getMessage());
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply server/src/main/java/com/thoughtworks/go/server/service/PipelineHistoryService.java]: Refactored to set only if `canRun==false && result.getServerHealthState!=null`

File: config/config-api/src/main/java/com/thoughtworks/go/config/Approval.java [hunk 3]
```diff
@@ -34,6 +34,10 @@
     @SkipParameterResolution
     private String type = SUCCESS;
 
+    @ConfigAttribute(value = "allowOnlyOnSuccess")
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply config/config-api/src/main/java/com/thoughtworks/go/config/Approval.java]: https://github.com/gocd/gocd/issues/6294#issuecomment-503537504

File: server/src/main/java/com/thoughtworks/go/server/service/StageManualTriggerChecker.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                    StageResult previousStageResult = previousStage.getResult();
+                    if (previousStageResult != StageResult.Passed) {
+                        String errorMessage = String.format("Cannot schedule %s as the previous stage %s has %s!", stageName, previousStageName.toString(), previousStageResult.name());
+                        healthStateType = HealthStateType.forbidden();
+                        result.notAcceptable(errorMessage, healthStateType);
+                        return;
+                    }
+                }
+            }
+        }
```
Comment (ketan) [inline server/src/main/java/com/thoughtworks/go/server/service/StageManualTriggerChecker.java]: Some of this looks like duplicate code from `ScheduleService#checkIfPreviousStagePassed`, think of de-duplicating some of this?

File: api/api-shared-v9/src/main/java/com/thoughtworks/go/apiv9/admin/shared/exceptions/InvalidGoCipherTextRuntimeException.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.apiv9.admin.shared.exceptions;
+
+public class InvalidGoCipherTextRuntimeException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply api/api-shared-v9/src/main/java/com/thoughtworks/go/apiv9/admin/shared/exceptions/InvalidGoCipherTextRuntimeException.java]: Removed the class as it is no longer utilized

File: config/config-api/src/main/java/com/thoughtworks/go/config/Approval.java [hunk 6]
```diff
@@ -92,21 +99,28 @@ public boolean equals(Object o) {
...
-        return true;
+        return allowOnlyOnSuccess == approval.allowOnlyOnSuccess;
     }
 
     @Override
     public int hashCode() {
-        int result = authConfig.hashCode();
-        result = 31 * result + type.hashCode();
-        return result;
+        return Objects.hash(authConfig, type, allowOnlyOnSuccess, errors);
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply config/config-api/src/main/java/com/thoughtworks/go/config/Approval.java]: Done. Thnks fr the tip. :)

File: config/config-server/src/test/java/com/thoughtworks/go/config/MagicalGoConfigXmlLoaderTest.java [hunk 7]
```diff
@@ -4438,6 +4438,64 @@ void shouldLoadRulesConfigWhereActionAndTypeHasWildcardForSchemaVersion124() thr
...
+                        "      <approval type=\"manual\" allowOnlyOnSuccess=\"true\" /> "
+                        + "    <jobs>"
+                        + "      <job name=\"functional\">"
+                        + "      </job>"
+                        + "    </jobs>"
+                        + "  </stage>"
+                        + "</pipeline>"
+                        + "</pipelines>", 126);
+
+        CruiseConfig config = xmlLoader.loadConfigHolder(content).config;
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply config/config-server/src/test/java/com/thoughtworks/go/config/MagicalGoConfigXmlLoaderTest.java]: Refactored.

File: server/src/main/java/com/thoughtworks/go/server/presentation/models/PipelineHistoryJsonPresentationModel.java [hunk 8]
```diff
@@ -189,6 +189,9 @@ private List stageHistoryAsJson(PipelineInstanceModel pipelineInstanceModel, Sta
                     stageHistoryItem.getName(), stageHistoryItem.getCounter());
             jsonMap.put("stageLocator", encodeInUtf8(stageIdentifier.stageLocator()));
             jsonMap.put("getCanRun", Boolean.toString(stageHistoryItem.getCanRun()));
+            if (!stageHistoryItem.getCanRun()) {
+                jsonMap.put("getErrorMessage", stageHistoryItem.getErrorMessage());
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply server/src/main/java/com/thoughtworks/go/server/presentation/models/PipelineHistoryJsonPresentationModel.java]: Updated to `errorMessage`

File: server/webapp/WEB-INF/rails/app/views/pipelines/_stage_trigger_gate.html.erb [hunk 9]
```diff
@@ -1,23 +1,25 @@
 <%
-title = if com.thoughtworks.go.util.GoConstants::DEFAULT_APPROVED_BY == scope[:stage_in_status_bar].getApprovedBy()
-    "Automatically approved"
-elsif scope[:stage_in_status_bar].getApprovedBy().blank?
-    "Awaiting approval"
-else
-    "Approved by #{scope[:stage_in_status_bar].getApprovedBy()}"
-end
+  title = if scope[:stage_in_status_bar].canRun() === false && scope[:stage_in_status_bar].getErrorMessage() != nil
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply server/webapp/WEB-INF/rails/app/views/pipelines/_stage_trigger_gate.html.erb]: Updated.

File: server/src/main/java/com/thoughtworks/go/server/service/ScheduleService.java [hunk 10]
```diff
@@ -540,6 +552,15 @@ public boolean canRun(PipelineIdentifier pipelineIdentifier, String stageName, S
         return result.getServerHealthState().isSuccess() && hasPreviousStageBeenScheduled;
     }
 
+    public boolean canRun(PipelineIdentifier pipelineIdentifier, String stageName, String username, boolean hasPreviousStageBeenScheduled, ServerHealthStateOperationResult result) {
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply server/src/main/java/com/thoughtworks/go/server/service/ScheduleService.java]: Refactored

File: server/src/main/java/com/thoughtworks/go/server/service/StageManualTriggerChecker.java (hunk 4 above)
Comment (kritika-singh3) [reply server/src/main/java/com/thoughtworks/go/server/service/StageManualTriggerChecker.java]: Done. Moved the logic to `SchedulingCheckerService.shouldAllowSchedulingStage`

File: server/src/test-fast/java/com/thoughtworks/go/server/service/StageManualTriggerCheckerTest.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    private StageManualTriggerChecker checker;
+    private String stageName;
+    private OperationResult result;
+    private int pipelineCounter;
+
+    @BeforeEach
+    void setUp() throws Exception {
+        schedulingCheckerService = mock(SchedulingCheckerService.class);
+        pipelineService = mock(PipelineService.class);
+        pipeline = mock(Pipeline.class);
```
Comment (ketan) [inline server/src/test-fast/java/com/thoughtworks/go/server/service/StageManualTriggerCheckerTest.java]: Why do you want to mock a POJO?

Comment (kritika-singh3) [reply server/src/test-fast/java/com/thoughtworks/go/server/service/StageManualTriggerCheckerTest.java]: It can be removed. Will make that change.

**Pull request merged**